pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+		/// specified for trade, or amount to be fulfilled.
+		InsufficientAssetFunds,
+		/// Error when an account does not have enough funds in chains reserve
+		/// currency to place order.
+		InsufficientReserveFunds,
+		/// Error when account tries to buy an invalid amount of currency --
+		/// currently `0`.
+		InvalidBuyAmount,
+		/// Error when an account specifies an invalid buy price -- currently
+		/// `0`.
+		InvalidMinPrice,
+		/// Error when an order is placed with a currency that is not in the
+		/// asset registry.
+		InvalidAssetId,
+		/// Error when an operation is attempted on an order id that is not in
+		/// storage.
+		OrderNotFound,
+		/// Error when a user attempts an action on an order they are not
+		/// authorised to perform, such as cancelling another accounts order.
+		Unauthorised,
+		/// Error when unable to convert fee balance to asset balance when asset
+		/// out matches fee currency
+		BalanceConversionErr,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		/// Create an order, with the minimum fulfillment amount set to the buy
+		/// amount, as the first iteration will not have partial fulfillment
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::Weights::create_order_v1())]
+		pub fn create_order_v1(
+			origin: OriginFor<T>,
+			asset_in: T::AssetCurrencyId,
+			asset_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			price: T::ForeignCurrencyBalance,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			Self::place_order(
+				account_id, asset_in, asset_out, buy_amount, price, buy_amount,
+			)?;
+			Ok(())
+		}
+
+		///  Cancel an existing order that had been created by calling account.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::Weights::user_cancel_order())]
+		pub fn user_cancel_order(
+			origin: OriginFor<T>,
+			order_id: T::OrderIdNonce,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			// verify order matches account
+			// UserOrders using Resultquery, if signed account
+			// does not match user for order id, we will get an Err Result
+			let order = <Orders<T>>::get(order_id)?;
+			if account_id == order.placing_account {
+				Self::cancel_order(order_id)?;
+				Ok(())
+			} else {
+				Err(DispatchError::from(Error::<T>::Unauthorised))
+			}
+		}
+
+		/// Fill an existing order, fulfilling the entire order.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::Weights::fill_order_full())]
+		pub fn fill_order_full(origin: OriginFor<T>, order_id: T::OrderIdNonce) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			let order = <Orders<T>>::get(order_id)?;
+			// maybe move to ensure if we don't need these later
+			// might need decimals from currency, but should hopefully be able to use FP
+			// price/amounts from FP balance
+
+			let sell_amount = order.buy_amount.ensure_mul(order.price)?;
```

Comment:
IIUC we are missing decimal conversion here. WDYT?

Reply:
Yes... I think you're right. If used `Balance` here have different decimal representation, we should unify them.

Reply:
Yeah, I think you're right, as we're dealing with the overall balances and not the adjusted precision from the registry.

Reply:
Regarding the conversions, in the Oracle PR, I add a module to perform that kind of conversions. Maybe you can extend that module @thea-leake 
 with a `balance_to_balance()` conversion: https://github.com/centrifuge/centrifuge-chain/pull/1474/files#diff-833d286c0e9655fa1b12da1bceabc7ff8988d1c74b7149be553320af61b78b54
 
Hopefully this PR is merged today or tomorrow